File: problems/binary-search-tree-check/binary-search-tree-check.py
class Tree(object):

    def __init__(self, val, left=None, right=None):
        self.val, self.left, self.right = val, left, right


def is_search_tree(tree, minval=float('-inf'), maxval=float('inf')):
    if not tree:
        # Empty
        return True
    if not (minval < tree.val < maxval):
        return False
    return  is_search_tree(tree.left, minval, tree.val) and \
        is_search_tree(tree.right, tree.val, maxval)

t1 = Tree(15,
          Tree(12,
               None,
               Tree(13)),
          Tree(22,
               Tree(18),
               Tree(100))
          )

t2 = Tree(15,
          Tree(18),
          None
          )

assert(is_search_tree(t1) == True)
assert(is_search_tree(t2) == False)

File: problems/binary-search-tree-check/binary_search_tree_check.rs
struct Tree {
    root: f64,
    left: Option<Box<Tree>>,
    right: Option<Box<Tree>>,
}

fn is_binary_search_tree(t: &Option<Box<Tree>>, min: f64, max: f64) -> bool {
    match *t {
        Some(ref t) => {
            if min < t.root && t.root < max {
                return is_binary_search_tree(&t.left, min, t.root) &&
                       is_binary_search_tree(&t.right, t.root, max);
            }
            false
        }
        None => true, // An empty tree is a balanced tree :-)
    }
}

fn main() {
    let t0 = Tree {
        root: 1.0,
        left: None,
        right: None,
    };
    assert_eq!(is_binary_search_tree(&Some(Box::new(t0)), -std::f64::INFINITY, std::f64::INFINITY),
               true);

    let t1 = Tree {
        root: 15.0,
        left: Some(Box::new(Tree {
            root: 12.0,
            left: None,
            right: Some(Box::new(Tree {
                root: 13.0,
                left: None,
                right: None,
            })),
        })),
        right: Some(Box::new(Tree {
            root: 22.0,
            left: Some(Box::new(Tree {
                root: 18.0,
                left: None,
                right: None,
            })),
            right: Some(Box::new(Tree {
                root: 100.0,
                left: None,
                right: None,
            })),
        })),
    };
    assert_eq!(true,
               is_binary_search_tree(&Some(Box::new(t1)), -std::f64::INFINITY, std::f64::INFINITY));

    let t2 = Tree {
        root: 15.0,
        left: Some(Box::new(Tree {
            root: 18.0,
            left: None,
            right: None,
        })),
        right: None,
    };

    assert_eq!(false,
               is_binary_search_tree(&Some(Box::new(t2)), -std::f64::INFINITY, std::f64::INFINITY));
}
